internal/provider/host_resource_retry.go

```diff
@@ -0,0 +1,146 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
+)
+
+// HostOperationFunc defines the signature for operations that can be retried
+type HostOperationFunc func() ([]byte, diag.Diagnostics, int)
+
+// HostRetryConfig contains configuration for retrying host operations
+type HostRetryConfig struct {
+	stateConf          *retry.StateChangeConf
+	operationName      string
+	operation          HostOperationFunc
+	successStatusCodes []int
+	ctx                context.Context
+}
+
+// Retry state constants
+const (
+	hostRetryStateRetrying = "retrying"
+	hostRetryStateSuccess  = "success"
+)
+
+// Retry timing constants
+const (
+	maxTimeoutSeconds = 30              // Maximum wait between retries (seconds)
+	minTimeoutSeconds = 5               // Minimum wait between retries (seconds)
+	delaySeconds      = 2               // Initial delay before first retry (seconds)
+	defaultBuffer     = 1 * time.Minute // Default buffer to leave between context deadline and timeout
+)
+
+// CalculateTimeout returns the retry timeout in seconds, which is 1 minute less than the context timeout.
+func CalculateTimeout(operationTimeoutSec int) int {
+	// Default fallback timeout in seconds
+	var timeout int
+
+	// If the deadline has already passed, use the minimum timeout
+	if operationTimeoutSec <= 0 {
+		return minTimeoutSeconds
+	}
+
+	// Use 80% of the remaining time for the timeout
+	calculatedTimeoutSeconds := operationTimeoutSec - int(defaultBuffer.Seconds())
+
+	// Ensure the timeout is at least the minimum viable timeout
+	if calculatedTimeoutSeconds < minTimeoutSeconds {
+		timeout = minTimeoutSeconds
+	} else {
+		// Return the timeout as an integer, truncating any fractions of a second
+		timeout = calculatedTimeoutSeconds
+	}
+
+	return timeout
+}
```

Comment:
Lets not do this calculation for now, 
Instead lets keep this as a flat 20 minutes,

I will make a story to expose timeout configuration to the user and we can work on something more sophisticated later 😄